exploration/beauty-contest.md

```diff
@@ -0,0 +1,316 @@
+
+The sigil `#` was chosen because `#define` type constructs are fairly common.
+Introduces `[`/`]` for keys.
+
+```
+#input {$var :function option=value}
+Hello {$var}
+
+#input {$var :function option=value}
+#local $foo = {$bar :function option=value}
+Hello {$var}, you have a {$foo}
+
+#match {$foo} {$bar}
+#when[foo bar] Hello {$foo} you have a {$var}
+#when[  *   *] {$foo} hello you have a {$var}
+
+#match {$foo :function option=value} {$bar :function option=value}
+#when [a b] {{  {$foo} is {$bar}  }}
+#when [x y] {||}  {$foo} is {$bar}  {||}
+#when [* *] {|  |}{$foo} is {$bar}{|  |}
+
+#{input $var :function option=value}#local $foo = {$bar :function option=value}Hello {$var}, you have a {$foo}
+
+#match {$foo} {$bar}#when[foo bar] Hello {$foo} you have a {$var}#when[* *] {$foo} hello you have a {$var}
+
+#match {$foo :function option=value} {$bar :function option=value}#when [a b] {{  {$foo} is {$bar}  }} #when [x y] {||}  {$foo} is {$bar}  {||}#when [* *] {|  |}{$foo} is {$bar}{|  |}
+```
+
+## 4. Reducing keywords
+
+Avoids keywords in favor of sigil based parsing.
+The theory here is that the syntactic sugar of `match` and `when` are nice the first time you use them
+but the benefit is lost or reduced after that.
+
+Uses double or paired sigils to reduce the need for escaping common characters (such as `?`)
+
+```
+#$var :function option=value
+Hello {$var}
+
+#$var :function option=value
+#$foo = {$bar :function option=value}
+Hello {$var}, you have a {$foo}
+
+?? {$foo} {$bar}
+::[ foo bar] Hello {$foo} you have a {$var}
+::[ *     *] {$foo} hello you have a {$var}
+
+?? {$foo :function option=value} {$bar :function option=value}
+::[a b] {{  {$foo} is {$bar}  }}
+::[x y] {{  {$foo} is {$bar}  }}
+::[* *] {|  |}{$foo} is {$bar}{|  |}
+
+#$var :function option=value}#$foo = {$bar :function option=value}Hello {$var}, you have a {$foo}
+
+??{$foo}{$bar}::[foo bar] Hello {$foo} you have a {$var}::[* *] {$foo} hello you have a {$var}
+
+??{$foo :function option=value}{$bar :function option=value}::[a b] {  {$foo} is {$bar}  }::[x y] {  {$foo} is {$bar}  }::[* *] {|  |}{$foo} is {$bar}{|  |}
+```
+
+## 5. Use "blocks" for declarations and body
+
+Use code-mode "blocks" to introduce code. _(This section has an additional example)_
+
+The theory here is that a "declarations block" would be a natural add-on and it doesn't
+require additional typing for each declaration.
+
+Note too that this syntax could be extended to allow other types of blocks,
+such as comments or different types of statement.
+
+```
+{#
+  input {$var :function option=value}
+}
+Hello {$var}
+
+// might be more natural as:
+{#input {$var :function option=value}}
+Hello {$var}
+
+{#
+   input {$var :function option=value}
+   local $foo = {$bar :function option=value}
+}
+Hello {$var}, you have a {$foo}
+
+{#
+  match {$foo} {$bar}
+  [ foo bar] Hello {$foo} you have a {$var}
+  [ *     *] {$foo} hello you have a {$var}
+}
+
+{#
+   input $foo :function option=value
+}{
```

Comment:
The original theory was: "declaration block" and "body block". The body block only needs to be bracketed when it contains code (i.e. a `match`). There's no reason to have blocks if they aren't separate (that would just be option 2 then)

Reply:
Ah, ok. Then I think lines 257 and 259 may need fixing, because together with this we've got all of:
- `{ match`
- `{#match`
- `match`

I can't tell which of these three is the proposed syntax here.